### DISCUSSION TITLE
How to use multiple connections to connect to existing database in memory mode in 0.7.1?

### DISCUSSION THREAD

**Discussion by yxmm-wxe** [Giorgi/DuckDB.NET#85] [General]
In the latest code, I found the commit [Isolate multiple in-memory database from each other](https://github.com/Giorgi/DuckDB.NET/commit/f7cfda7c63654d88aad61e97a67e7e3cdde2fca5).

So when I get an in-memory connection every time, it will be connected to a new database.

How to use multiple connections to connect to the existing database in memory mode? Any plans to expand this functionality?

**Comment by Giorgi** [comment]
@yxmm-wxe SQLite supports sharing in-memory databases between connections: [In-Memory Databases](https://www.sqlite.org/inmemorydb.html) but I don't see anything like that in DuckDB docs. I asked in [DuckDB Discord](https://discord.com/channels/909674491309850675/921073327009853451/1086377779823652886), let's see what answer I get.

**Reply by yxmm-wxe** [reply]
@Giorgi Thanks! Maybe I didn't express myself clearly. Before the commit mentioned above, refer to [ConnectionManager](https://github.com/Giorgi/DuckDB.NET/blob/main/DuckDB.NET.Data/Internal/ConnectionManager.cs), first time it will get a connection via `NativeMethods.Startup.DuckDBOpen`. Then second time it will get another connection via `NativeMethods.Startup.DuckDBConnect`.  As my understanding, it is sharing in-memory databases between connections.

After the commit [Isolate multiple in-memory database from each other](https://github.com/Giorgi/DuckDB.NET/commit/f7cfda7c63654d88aad61e97a67e7e3cdde2fca5), I'd like to know how to share in-memory databases between connections like before.

**Reply by Giorgi** [reply]
It's not currently possible with DuckDB.NET I think it isn't supported in any other client library for DuckDB either. I can probably implement it on top of the current functionality but I can't promise that I'll be able to do it in near future.

Out of curiosity, in which scenario do you need it?

**Reply by yxmm-wxe** [reply]
I found that it was supported in JDBC, please refer to [DuckDB JDBC duplicate method](https://duckdb.org/docs/archive/0.6.1/api/java#:~:text=The%20JDBC%20DriverManager,method%20like%20so%3A).

I'm just learning to use it personally, and I've tried to use DuckDB JDBC before. So just curious if there are plans to expand this functionality?

**Reply by Giorgi** [reply]
JDBC supports it because JDBC driver is built by the core DuckDB team while this project is built mostly by only me in my spare time. As I said already I can't promise that I will do it in near future.

**Reply by yxmm-wxe** [reply]
Got it. Thanks for your contribution. I am also willing to try to contribute this functionality in my spare time.

**Reply by yxmm-wxe** [reply]
@Giorgi I am very sorry to ask a low-level question.

I'd like to build and run a test in VS on Windows. The command `dotnet build -c Debug /p:BuildType=Full` is executed. But when I run a test in VS, it always prompts the following error. While it does have `duckdb.dll` in `DuckDB.NET\bin\Debug\net6.0\runtimes\win-x64` and `DuckDB.NET\obj\runtimes\win-x64\native`.
```
System.DllNotFoundException : Unable to load DLL 'duckdb' or one of its dependencies: The specified module could not be found. (0x8007007E)
```

Are there any missing steps?

**Reply by Giorgi** [reply]
Are you running tests in Debug mode? There is ModuleInit class the loads duckdb before running tests. Try putting a breakpoint there and debugging tests to see where it is looking for the duckdb.dll.

**Reply by yxmm-wxe** [reply]
Thanks! It works.

**Comment by Giorgi** [comment]
I created a new issue to track this functionality: [Support shared in-memory databases](https://github.com/Giorgi/DuckDB.NET/issues/86)

**Reply by yxmm-wxe** [reply]
Already implemented in #88.

It can use the connection string `DataSource=:memory:?cache=shared` to get and use the shared in-memory database.

It can also use the connection string `DataSource=:memory:` to get and use an isolated in-memory database and use `connection.Duplicate()` to get another connection for the existing isolated in-memory database.
